src/IO/S3/Client.cpp

```diff
@@ -0,0 +1,400 @@
+{
+    return wrapped_strategy->RequestBookkeeping(httpResponseOutcome);
+}
+
+void Client::RetryStrategy::RequestBookkeeping(const Aws::Client::HttpResponseOutcome& httpResponseOutcome, const Aws::Client::AWSError<Aws::Client::CoreErrors>& lastError)
+{
+    return wrapped_strategy->RequestBookkeeping(httpResponseOutcome, lastError);
+}
+
+bool Client::checkIfWrongRegionDefined(const std::string & bucket, const Aws::S3::S3Error & error, std::string & region) const
+{
+    if (detect_region)
+        return false;
+
+    if (error.GetResponseCode() == Aws::Http::HttpResponseCode::BAD_REQUEST && error.GetExceptionName() == "AuthorizationHeaderMalformed")
+    {
+        region = GetErrorMarshaller()->ExtractRegion(error);
+
+        if (region.empty())
+            region = getRegionForBucket(bucket, /*force_detect*/ true);
+
+        assert(!explicit_region.empty());
+        if (region == explicit_region)
+            return false;
+
+        insertRegionOverride(bucket, region);
+        return true;
+    }
+
+    return false;
+}
+
+void Client::insertRegionOverride(const std::string & bucket, const std::string & region) const
+{
+    std::lock_guard lock(cache->region_cache_mutex);
+    auto [it, inserted] = cache->region_for_bucket_cache.emplace(bucket, region);
+    if (inserted)
+        LOG_INFO(log, "Detected different region ('{}') for bucket {} than the one defined ('{}')", region, bucket, explicit_region);
+}
+
+Model::HeadObjectOutcome Client::HeadObject(const HeadObjectRequest & request) const
+{
+    const auto & bucket = request.GetBucket();
+
+    if (auto region = getRegionForBucket(bucket); !region.empty())
+    {
+        if (!detect_region)
+            LOG_INFO(log, "Using region override {} for bucket {}", region, bucket);
+        request.overrideRegion(std::move(region));
+    }
+
+    if (auto uri = getURIForBucket(bucket); uri.has_value())
+        request.overrideURI(std::move(*uri));
+
+    auto result = Aws::S3::S3Client::HeadObject(request);
+    if (result.IsSuccess())
+        return result;
+
+    const auto & error = result.GetError();
+
+    std::string new_region;
+    if (checkIfWrongRegionDefined(bucket, error, new_region))
+    {
+        request.overrideRegion(new_region);
+        return HeadObject(request);
+    }
+
+    if (error.GetResponseCode() != Aws::Http::HttpResponseCode::MOVED_PERMANENTLY)
+        return result;
+
+    // maybe we detect a correct region
+    if (!detect_region)
+    {
+        if (auto region = GetErrorMarshaller()->ExtractRegion(error); !region.empty() && region != explicit_region)
+        {
+            request.overrideRegion(region);
+            insertRegionOverride(bucket, region);
+        }
+    }
+
+    auto bucket_uri = getURIForBucket(bucket);
+    if (!bucket_uri)
+    {
+        if (auto maybe_error = updateURIForBucketForHead(bucket); maybe_error.has_value())
```

Comment:
I don't think we should make our code that complicated. If `HeadBucket` and the `HeadObject`'s response header both didn't help us to get a correct region then `ListObjects` will hardly help.

Reply:
It will help us if we got 301 but we couldn't get the location because HEAD doesn't have a response body.